azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java

```diff
@@ -1053,4 +1056,63 @@ public void shutdown() {
       this.scheduler.shutdownNow();
     }
   }
+
+  private class PollingCriteria {
+
+    private final Props azkabanProps;
+    private final SystemMemoryInfo memInfo = SERVICE_PROVIDER.getInstance(SystemMemoryInfo.class);
+    private final OsCpuUtil cpuUtil = SERVICE_PROVIDER.getInstance(OsCpuUtil.class);
+
+
+    public boolean shouldPoll() {
+      if (satisfiesFlowThreadsAvailableCriteria() && satisfiesFreeMemoryCriteria()
+          && satisfiesCpuUtilizationCriteria()) {
+        return true;
+      }
+      return false;
+    }
+
+    public PollingCriteria(final Props azkabanProps) {
+      this.azkabanProps = azkabanProps;
+    }
+
+    private boolean satisfiesFlowThreadsAvailableCriteria() {
+      final int flowThreadsAvailableConfig = this.azkabanProps.
+          getInt(ConfigurationKeys.AZKABAN_POLLING_CRITERIA_FLOW_THREADS_AVAILABLE, 0);
+
+      // allow polling if not present or configured with invalid value
+      if (flowThreadsAvailableConfig <= 0) {
+        return true;
+      }
+      final int remainingFlowThreads = FlowRunnerManager.this.getMaxNumRunningFlows() -
+          FlowRunnerManager.this.getNumRunningFlows();
```

Comment:
Do you think we can add a log message here when `remainingFlowThreads` <= 0 so that we know this check kicks in?

Reply:
yes, I'll do it.